File: DSA/RiddhiSaxena_590022853/Day_18/question1.py
class Solution:
    def rotateRight(self, head, k):
        if not head or not head.next or k == 0:
            return head

        # Find the length of the list and the last node
        length = 1
        tail = head

        while tail.next:
            tail = tail.next
            length += 1

        # Reduce unnecessary rotations
        k = k % length
        if k == 0:
            return head

        # Make the list circular
        tail.next = head

        # Find the new tail
        steps = length - k - 1
        new_tail = head

        for _ in range(steps):
            new_tail = new_tail.next

        # Set the new head
        new_head = new_tail.next

        # Break the circle
        new_tail.next = None

        return new_head

File: DSA/SanyamGambhir_590016388/Day_17/Question_1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 * int val;
 * ListNode *next;
 * ListNode() : val(0), next(nullptr) {}
 * ListNode(int x) : val(x), next(nullptr) {}
 * ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        // 1. Handle edge cases: empty list, single node, or no rotation
        if (!head || !head->next || k == 0) {
            return head;
        }
        
        // 2. Compute the length of the list and find the tail node
        ListNode* last = head;
        int length = 1;
        while (last->next) {
            last = last->next;
            length++;
        }
        
        // 3. Normalize k to avoid redundant full rotations
        k = k % length;
        if (k == 0) {
            return head;
        }
        
        // 4. Connect the tail to the head to make it circular
        last->next = head;
        
        // 5. Find the new tail: located (length - k - 1) steps from the head
        int stepsToNewTail = length - k - 1;
        ListNode* newTail = head;
        for (int i = 0; i < stepsToNewTail; i++) {
            newTail = newTail->next;
        }
        
        // 6. Establish the new head and break the circular bond
        ListNode* newHead = newTail->next;
        newTail->next = nullptr;
        
        return newHead;
    }
};